nodl_python/nodl/_parsing/_v1/_parsing.py

```diff
@@ -0,0 +1,114 @@
+
+    server = get_bool_attribute(element, 'server')
+    client = get_bool_attribute(element, 'client')
+    if not (server or client):
+        raise errors.AmbiguousActionInterfaceError(element)
+
+    policy = _parse_qos(element.find('qos'))
+
+    return Action(name=name, action_type=action_type, server=server, client=client, qos=policy)
+
+
+def _parse_parameter(element: etree._Element) -> Parameter:
+    """Parse a NoDL parameter from an xml element."""
+    return Parameter(name=element.get('name'), parameter_type=element.get('type'))
+
+
+def _parse_service(element: etree._Element) -> Service:
+    """Parse a NoDL service from an xml element."""
+    name = element.get('name')
+    service_type = element.get('type')
+
+    server = get_bool_attribute(element, 'server')
+    client = get_bool_attribute(element, 'client')
+    if not (server or client):
+        raise errors.AmbiguousServiceInterfaceError(element)
+
+    policy = _parse_qos(element.find('qos'))
+
+    return Service(name=name, service_type=service_type, server=server, client=client, qos=policy,)
+
+
+def _parse_topic(element: etree._Element) -> Topic:
+    """Parse a NoDL topic from an xml element."""
+    name = element.get('name')
+    message_type = element.get('type')
+
+    publisher = get_bool_attribute(element, 'publisher')
+    subscription = get_bool_attribute(element, 'subscription')
+    if not (publisher or subscription):
+        raise errors.AmbiguousTopicInterfaceError(element)
+
+    policy = _parse_qos(element.find('qos'))
+
+    return Topic(
+        name=name,
+        message_type=message_type,
+        publisher=publisher,
+        subscription=subscription,
+        qos=policy,
+    )
+
+
+def _parse_nodes(interface: etree._Element) -> List[Node]:
+    """Parse the nodes contained in an interface element and return a list."""
+    node_elements = [child for child in interface if child.tag == 'node']
+    return [_parse_node(node) for node in node_elements]
+
+
+def _parse_node(node: etree._Element) -> Node:
+    """Parse a NoDL node and all the elements it contains from an xml element."""
+    name = node.attrib['name']
+
+    actions = []
+    parameters = []
+    services = []
+    topics = []
+
+    for child in node:
+        if child.tag == 'action':
+            actions.append(_parse_action(child))
+        if child.tag == 'parameter':
+            parameters.append(_parse_parameter(child))
+        if child.tag == 'service':
+            services.append(_parse_service(child))
+        if child.tag == 'topic':
+            topics.append(_parse_topic(child))
+    return Node(
+        name=name, actions=actions, parameters=parameters, services=services, topics=topics
+    )
+
+
+def _validate_and_parse(interface: etree._Element) -> List[Node]:
+    """"""
+    try:
+        v1_schema().assertValid(interface)
+    except etree.DocumentInvalid as e:
+        raise errors.InvalidNoDLError(e.error_log[0].message, e) from e
```

Comment:
Again, passing strings to an exception means it's not specific enough (or just needs to be refactored). You're passing `e` anyway, it should be able to pull the log out itself. And in fact... the error already does support that.

Reply:
36f9562